includes/utils/class-amp-validation-utils.php

```diff
@@ -1791,4 +1878,60 @@ public static function get_recheck_link( $post, $redirect_url, $recheck_url = nu
 		);
 	}
 
+	/**
+	 * Enqueues the block validation script.
+	 *
+	 * @return void
+	 */
+	public static function enqueue_block_validation() {
+		$slug = 'amp-block-validation';
+
+		wp_enqueue_script(
+			$slug,
+			amp_get_asset_url( "js/{$slug}.js" ),
+			array( 'jquery' ),
+			AMP__VERSION,
+			true
+		);
+
+		$data = wp_json_encode( array(
+			'i18n'     => array(
+				/* translators: %s: the name of the block */
+				'notice' => __( 'The %s block above has invalid AMP', 'amp' ),
+			),
+		) );
+		wp_add_inline_script( $slug, sprintf( 'ampBlockValidation.boot( %s );', $data ) );
+	}
+
+	/**
+	 * Adds fields to the REST API responses, in order to display validation errors.
+	 *
+	 * @return void
+	 */
+	public static function add_rest_api_fields() {
+		register_rest_field(
+			array( 'post', 'page' ),
+			self::REST_FIELD_NAME,
+			array(
+				'get_callback' => array( __CLASS__, 'rest_field_amp_validation' ),
+				'schema'       => array(
+					'description' => __( 'AMP validation results', 'amp' ),
+					'type'        => 'object',
+				),
+			)
+		);
+	}
+
+	/**
+	 * Adds a field to the REST API responses to display the validation status.
+	 *
+	 * @param array  $post_data  Data for the post.
+	 * @param string $field_name The name of the field to add.
+	 * @return array|null $validation_data Validation data if it's available, or null.
+	 */
+	public static function rest_field_amp_validation( $post_data, $field_name ) {
+		$validation_post = self::get_validation_status_post( $post_data['link'] );
```

Comment:
Excellent re-use of this function. 

One key thing is that instead of `$post_data['link']` it should do `amp_get_permalink( $post_data['id'] )` so that it can work in paired mode.

Idea: If this post is empty or the post type is not viewable, should we take the post content and try to validate it? In other words, this:

https://github.com/Automattic/amp-wp/blob/0bd1d58451d7a78b125e8eac2f1b7d02ebfb124a/includes/utils/class-amp-validation-utils.php#L549-L560

Reply:
Good idea, @westonruter. [This commit](https://github.com/Automattic/amp-wp/pull/1019/commits/ed4b03ba8a01dd276cc79378456d100a84088552) abstracts that snippet above into [get_existing_validation_errors()](https://github.com/Automattic/amp-wp/pull/1019/commits/ed4b03ba8a01dd276cc79378456d100a84088552#diff-7d234989187d87a372e2aa6aa1299a71R636), and reuses it here. If there are no existing errors, it [validates the front-end](https://github.com/Automattic/amp-wp/pull/1019/commits/ed4b03ba8a01dd276cc79378456d100a84088552#diff-7d234989187d87a372e2aa6aa1299a71R1968). 

The only issue is that this delays the response if it validates several posts.